### DISCUSSION TITLE
Protecting ConcurrentHashMap with ReadWriteLock

### DISCUSSION THREAD

**Discussion by Montura** [apache/ozone#6644] [Ideas]
Example: 
- there is a class [NodeStateMap](https://github.com/apache/ozone/blob/master/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/states/NodeStateMap.java#L47).

It has 3 fields:
1. `ConcurrentHashMap<UUID, DatanodeInfo> nodeMap`
2. `ConcurrentHashMap<UUID, Set<ContainerID>> nodeToContainer`
3. `private final ReadWriteLock lock`

Both `read` and `write` operations in `NodeStateMap` are protected by `lock`.

There is a question:
- Why do we need to use `ConcurrentHashMap` here instead of `HashMap`, if all `methods` are protected?
- I'd like to made a patch after some discussion

**Comment by adoroszlai** [comment]
@nandakumar131 may remember, but on first look it seems plain `HashMap` would be OK.

**Comment by nandakumar131** [comment]
@Montura you are right.
As @adoroszlai also pointed out, we should be good with plain `HashMap` as we have lock to protect the access to the internal maps.

Please go ahead and file a Jira in Apache.

**Reply by Montura** [reply]
@nandakumar131, thanks! I'll do it and will make a PR
